python/tk_framework_desktopserver/shotgun/scripts/cache_commands.py

```diff
@@ -0,0 +1,185 @@
+import cPickle
+
+class _BootstrapLogHandler(logging.StreamHandler):
+    """
+    Manually flushes emitted records for js to pickup.
+    """
+
+    def emit(self, record):
+        """
+        Forwards the record back to to js via the engine communicator.
+
+        :param record: The record to log.
+        """
+        super(_BootstrapLogHandler, self).emit(record)
+
+        # always flush to ensure its seen by the js process
+        self.flush()
+
+def get_sgtk_logger(sgtk):
+    """
+    Sets up a log handler and logger.
+
+    :param sgtk: An sgtk module reference.
+
+    :returns: A logger and log handler.
+    """
+    # add a custom handler to the root logger so that all toolkit log messages
+    # are forwarded back to python via the communicator
+    bootstrap_log_formatter = logging.Formatter("[%(levelname)s]: %(message)s")
+    bootstrap_log_handler = _BootstrapLogHandler()
+    bootstrap_log_handler.setFormatter(bootstrap_log_formatter)
+    bootstrap_log_handler.setLevel(logging.DEBUG)
+
+    # now get a logger to use during bootstrap
+    sgtk_logger = sgtk.LogManager.get_logger("%s.%s" % ("tk-shotgun", "bootstrap"))
+    sgtk.LogManager().initialize_custom_handler(bootstrap_log_handler)
+
+    # allows for debugging to be turned on by the plugin build process
+    sgtk.LogManager().global_debug = True
+
+    # initializes the file where logging output will go
+    sgtk.LogManager().initialize_base_file_handler("tk-shotgun")
+    sgtk_logger.debug("Log dir: %s" % (sgtk.LogManager().log_folder))
+
+    return sgtk_logger, bootstrap_log_handler
+
+def cache(cache_file, data, base_configuration, hash_data, engine_name):
+    import sqlite3
+    import sgtk
+
+    entity = dict(type=data["entity_type"], id=data["entity_id"])
+    project_entity = dict(id=data["project_id"], type="Project")
+
+    # Setup the bootstrap manager.
+    toolkit_mgr = sgtk.bootstrap.ToolkitManager()
+    toolkit_mgr.plugin_id = "shotgun_toolkit_menu_caching"
+    toolkit_mgr.base_configuration = base_configuration
+
+    for pc in data["pipeline_configs"]:
+        pc["project"] = pc.get("project", project_entity)
+    import pprint
+    pprint.pprint(data["pipeline_configs"])
+    pcs = toolkit_mgr.sort_and_filter_configuration_entities(
+        project=project_entity,
+        entities=data["pipeline_configs"],
+    ) or [dict(id=None)]
+
+    for pc in pcs:
+        logger, log_handler = get_sgtk_logger(sgtk)
+        pprint.pprint(hash_data)
+        lookup_hash = hash_data[pc["id"]]["lookup_hash"]
+        contents_hash = hash_data[pc["id"]]["contents_hash"]
+
+        toolkit_mgr.pipeline_configuration = pc["id"]
+        engine = toolkit_mgr.bootstrap_engine(engine_name, entity=entity)
+        pc_descriptor = toolkit_mgr.resolve_descriptor(project_entity)
+
+        # Clean up the pre-bootstrap logger since we can use the engine's
+        # logger from here on out.
+        sgtk.LogManager().root_logger.removeHandler(log_handler)
```

Comment:
Is this because tk-shotgun always logs on the console? I think adding a comment to that effect would be nice.